### ISSUE TITLE
kube-proxy fails to retrieve services and endpoints in Vagrant/Virtualbox cluster

### ISSUE THREAD

**Issue by markllama** [kubernetes/kubernetes#7622]
Possibly related to https://github.com/GoogleCloudPlatform/kubernetes/pull/7303

On simply starting a Vagrant/Virtualbox cluster and examining the journalctl results for the kube-proxy startup on minion-1:

    [vagrant@kubernetes-minion-1 ~]$ journalctl --unit kube-proxy | head 
    -- Logs begin at Tue 2015-02-24 15:05:10 UTC, end at Fri 2015-05-01 14:59:21 UTC. --
    May 01 13:51:38 kubernetes-minion-1 systemd[1]: Starting Kubernetes Kube-Proxy Server...
    May 01 13:51:38 kubernetes-minion-1 systemd[1]: Started Kubernetes Kube-Proxy Server.
    May 01 13:51:38 kubernetes-minion-1 kube-proxy[3231]: I0501 13:51:38.155272    3231 server.go:86] Running in resource-only container "/kube-proxy"
    May 01 13:51:38 kubernetes-minion-1 kube-proxy[3231]: I0501 13:51:38.157258    3231 proxier.go:336] Setting Proxy IP to 10.0.2.15
    May 01 13:51:38 kubernetes-minion-1 kube-proxy[3231]: I0501 13:51:38.157285    3231 proxier.go:341] Initializing iptables
    May 01 13:51:38 kubernetes-minion-1 kube-proxy[3231]: E0501 13:51:38.211881    3231 api.go:180] Unable to load endpoints: the server has asked for the client to provide credentials (get endpoints)
    May 01 13:51:38 kubernetes-minion-1 kube-proxy[3231]: E0501 13:51:38.216027    3231 api.go:108] Unable to load services: the server has asked for the client to provide credentials (get services)
    May 01 13:52:58 kubernetes-minion-1 kube-proxy[3231]: E0501 13:52:58.077294    3231 api.go:180] Unable to load endpoints: the server has asked for the client to provide credentials (get endpoints)
    May 01 13:53:07 kubernetes-minion-1 kube-proxy[3231]: E0501 13:53:07.874594    3231 api.go:108] Unable to load services: the server has asked for the client to provide credentials (get services)

It appears that the kube-proxy initialization process is not authenticating to the server correctly:

    systemctl status kube-proxy | cat
    kube-proxy.service - Kubernetes Kube-Proxy Server
        Loaded: loaded (/usr/lib/systemd/system/kube-proxy.service; enabled)
       Active: active (running) since Fri 2015-05-01 13:51:38 UTC; 1h 16min ago
         Docs: https://github.com/GoogleCloudPlatform/kubernetes
     Main PID: 3231 (kube-proxy)
        CGroup: /system.slice/kube-proxy.service
                └─3231 /usr/local/bin/kube-proxy --master=https://10.245.1.2 --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2

    May 01 15:02:06 kubernetes-minion-1 kube-proxy[3231]: E0501 15:02:06.687631    3231 api.go:108] Unable to load services: the server has asked for the client to provide credentials (get services)
    May 01 15:02:52 kubernetes-minion-1 kube-proxy[3231]: E0501 15:02:52.509411    3231 api.go:180] Unable to load endpoints: the server has asked for the client to provide credentials (get endpoints)
    ....

````
sudo cat /var/lib/kube-proxy/kubeconfig
apiVersion: v1
kind: Config
users:
- name: kube-proxy
  user:
    token: UUalZjWmetMofOkoQNZhIqxzWA7X82mR
clusters:
- name: local
  cluster:
     insecure-skip-tls-verify: true
contexts:
- context:
    cluster: local
    user: kube-proxy
  name: service-account-context
current-context: service-account-context
````

**Comment by markllama** [comment]
This is possibly also related: https://github.com/GoogleCloudPlatform/kubernetes/issues/6164

````
cat /etc/sysconfig/kube-proxy 
DAEMON_ARGS=" --master=https://10.245.1.2 --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2"
````

I'm going to try port 7something and see if that changes anything.

**Comment by markllama** [comment]
switching master to port 708 results in this:

````
cat /etc/sysconfig/kube-proxy
DAEMON_ARGS=" --master=https://10.245.1.2:7080 --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2"

systemctl restart kube-proxy
systemctl status kube-proxy -l
kube-proxy.service - Kubernetes Kube-Proxy Server
   Loaded: loaded (/usr/lib/systemd/system/kube-proxy.service; enabled)
   Active: active (running) since Fri 2015-05-01 15:24:41 UTC; 14s ago
     Docs: https://github.com/GoogleCloudPlatform/kubernetes
 Main PID: 26978 (kube-proxy)
   CGroup: /system.slice/kube-proxy.service
           └─26978 /usr/local/bin/kube-proxy --master=https://10.245.1.2:7080 --kubeconfig=/var/lib/kube-proxy/kubeconfig --v=2

May 01 15:24:41 kubernetes-minion-1 systemd[1]: Started Kubernetes Kube-Proxy Server.
May 01 15:24:41 kubernetes-minion-1 kube-proxy[26978]: I0501 15:24:41.677577   26978 server.go:86] Running in resource-only container "/kube-proxy"
May 01 15:24:41 kubernetes-minion-1 kube-proxy[26978]: I0501 15:24:41.680722   26978 proxier.go:336] Setting Proxy IP to 10.0.2.15
May 01 15:24:41 kubernetes-minion-1 kube-proxy[26978]: I0501 15:24:41.680772   26978 proxier.go:341] Initializing iptables
May 01 15:24:41 kubernetes-minion-1 kube-proxy[26978]: E0501 15:24:41.721230   26978 api.go:108] Unable to load services: Get https://10.245.1.2:7080/api/v1beta3/services: tls: oversized record received with length 20527
May 01 15:24:41 kubernetes-minion-1 kube-proxy[26978]: E0501 15:24:41.721406   26978 api.go:180] Unable to load endpoints: Get https://10.245.1.2:7080/api/v1beta3/endpoints: tls: oversized record received with length 20527
````

Then IP tables (no services) looks like this:
sudo iptables -vL -n -t nat
Chain PREROUTING (policy ACCEPT 7 packets, 420 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    3   148 DOCKER     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL
  194 11624 KUBE-PORTALS-CONTAINER  all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 85 packets, 5100 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DOCKER     all  --  *      *       0.0.0.0/0           !127.0.0.0/8          ADDRTYPE match dst-type LOCAL
 2378  142K KUBE-PORTALS-HOST  all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain POSTROUTING (policy ACCEPT 85 packets, 5100 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  192 11520 MASQUERADE  all  --  *      !kbr0   10.246.1.0/24        0.0.0.0/0           
    0     0 MASQUERADE  all  --  *      !docker0  172.17.0.0/16        0.0.0.0/0           

Chain DOCKER (2 references)
 pkts bytes target     prot opt in     out     source               destination         

Chain KUBE-PORTALS-CONTAINER (1 references)
 pkts bytes target     prot opt in     out     source               destination         

Chain KUBE-PORTALS-HOST (1 references)
 pkts bytes target     prot opt in     out     source               destination         
````

**Comment by justinsb** [comment]
I was able to get past this by changing the --master argument to kube-proxy from --master=https://172.20.0.9 to --master=https://172.20.0.9:6443.

I believe the former (port 443) is actually nginx, which uses basic auth.  The latter (6443) is direct-to-kube-apiserver, which is what we want.

I think GCE works only because on GCE nginx is disabled (top.sls), and apiserver listens on 443 on GCE (kube-apiserver.manifest).

I'm going to look into fixing for AWS by doing the same thing as GCE: disable nginx, have apiserver listen on 443.  If that works, should be easy to do the same for vagrant.

**Comment by derekwaynecarr** [comment]
I think there is an issue open that says I need to do the same, just lacked time this week to disable nginx

**Comment by markllama** [comment]
PR to fix the Vagrant test cluster: https://github.com/GoogleCloudPlatform/kubernetes/pull/7791

**Comment by markllama** [comment]
I finally found the right hook I think:

````
find cluster -type f | xargs grep 6443 
cluster/vagrant/provision-minion.sh:  api_servers: '$(echo "$MASTER_IP" | sed -e "s/'/''/g"):6443'
cluster/saltbase/salt/kubelet/default:  {% set api_servers_with_port = api_servers + ":6443" -%}
cluster/saltbase/salt/kube-apiserver/kube-apiserver.manifest:{% set secure_port = "6443" -%}
````

We could either leave my patch which would set both in provision-minion.sh OR I can remove that change and add the kubelet string suffix pattern to the kube-proxy default template.

Preference?

**Comment by markllama** [comment]
Re-submitted PR to patch:
- https://github.com/GoogleCloudPlatform/kubernetes/pull/7791

This time I used the same mechanism that was used to do the same thing for kubelet.

**Comment by roberthbailey** [comment]
@markllama This was definitely broken by #7303, which switched everyone other than azure to port 443.

**Comment by roberthbailey** [comment]
From what I can tell, this was fixed by #7794.
